### DISCUSSION TITLE
mosdns与ssr plus、passwall等科学软件配合的原理是什么？

### DISCUSSION THREAD

**Discussion by yyysuo** [IrineSistiana/mosdns#470] [General]
passwall等软件本身是有dns分流的，也是根据geoip和geosite规则来分流的，网上查资料说获取的dns结果还会写入ipset，那么用mosdns不走passwall控制的dnsmasq，还能写入ipset吗？如果mosdns的分流规则与科学软件的分流规则不一致，比如某域名在mosdns中走了国外解析，但是科学软件中走了国内，那么就是国内直连，反之就是mosdns走了国内解析，解析出国内ip，但是在科学软件中是国外的，那么就会出国绕一圈。

另外如果仅把mosdns作为一个dns服务，手工去设置dns转发的话，ssr plus和passwall应该如何设置？

**Comment by qingtian110** [comment]
Q：那么用mosdns不走passwall控制的dnsmasq，还能写入ipset吗？
A：能，但没必要，内存缓存几十万条应答资源占用非常小且效率高，如果一定需要，mosdns也支持把应答写入ipset 、nftables

Q：什么什么绕了一圈
A：假如你使用了科学自带的dns插件，完全无需再使用mosdns；如果你使用了mosdns，你应该把科学插件的上游dns设置为 tcp/udp方式的 127.0.0.1:mosdns监听端口

Q：ssr plus和passwall应该如何设置？
A：只是建议

ssrp：使用本机端口为5335的DNS服务
passwall： 过滤模式：通过UDP请求DNS  远程DNS：127.0.0.1:5335

**Reply by yyysuo** [reply]
Q：那么用mosdns不走passwall控制的dnsmasq，还能写入ipset吗？
A：能，但没必要，内存缓存几十万条应答资源占用非常小且效率高，如果一定需要，mosdns也支持把应答写入ipset 、nftables

问：不了解科学软件的内部处理逻辑，写入ipset是否是科学软件必要的呢？

Q：什么什么绕了一圈
A：假如你使用了科学自带的dns插件，完全无需再使用mosdns；如果你使用了mosdns，你应该把科学插件的上游dns设置为 tcp/udp方式的 127.0.0.1:mosdns监听端口
问：科学软件和它的dns方案一般是一体的，ssrp和passwall貌似不可以单独关闭，opeclash好像更不可以。

Q：ssr plus和passwall应该如何设置？
A：只是建议

ssrp：使用本机端口为5335的DNS服务
passwall： 过滤模式：通过UDP请求DNS 远程DNS：127.0.0.1:5335
问：这只处理了国外域名的dns，国内的域名是不是就无法处理了？如果在dhcp转发中把国内dns也转发到modns，等于是科学软件分流一次，然后分流完再汇总，mosdns再分一次？

**Comment by qingtian110** [comment]
当然不是这样，如果你想用mosdns 在软路由上，我建议你直接使用[openwrt 版本](https://github.com/sbwml/luci-app-mosdns#how-to-install-prebuilt-packages-openwrt-18--19--21--22--master)，这个你无需自己配置mosdns，它默认就是分流解析。

另外你上面的疑问 `这只处理了国外域名的dns，国内的域名是不是就无法处理了？如果在dhcp转发中把国内dns也转发到modns，等于是科学软件分流一次，然后分流完再汇总，mosdns再分一次？` 

A：国内域名它会自动使用WAN口的dns地址进行解析，国外域名则使用远程DNS解析，而这两个dns你都可以自定义设置。

mosdns 在openwrt 上它直接接管了dnsmasq 的解析操作（也就是你说的dhcp），这只要开启mosdns的转发功能就行
![1](https://user-images.githubusercontent.com/87856890/186450015-a2351129-bd89-4383-8a7e-100b0f89cc37.png)



然而你上面问的两种，如果你使用了mosdns ，里应按照下面图片这样设置科学插件，这样全局域名解析全部由mosdns接管，科学插件带的dns也不会独立去运行

passwall
![11](https://user-images.githubusercontent.com/87856890/186451697-d95b4e05-94d2-4604-b4da-fd2d7309e372.png)


ssr
![22](https://user-images.githubusercontent.com/87856890/186451777-dbfbc6ec-e951-4cf4-935a-5f23d59e6a72.png)

**Reply by yyysuo** [reply]
然而你上面问的两种，如果你使用了mosdns ，里应按照下面图片这样设置科学插件，这样全局域名解析全部由mosdns接管，科学插件带的dns也不会独立去运行

回应：以我的理解，咱俩的说法并没有冲突的地方。以ssr为例，按你的设置的话，国外域名部分是ssrp通过luci中的远程dns地址向dnsmqsq添加域名规则，转发dns到本机的5335端口。国内部分ssr plus没做处理，默认dnsmasq转发到wan口的运营商dns。启用mosdns之后，dhcp的dns转发会自动填上127.0.0.1#5335吧？国内部分就转发到mosdns了。这个过程中，国内国外dns的区分，就是ssrp读取luci高级设置中设定的分流规则进行分流的体现，科学软件的分流是切实存在的。科学插件并没有带dns，而是通过dnsmasq的dns服务和规则进行分流的。
而后才有ssrp分流之后的两种dns又汇总到mosdns中，又根据mosdns的分流规则进行分流。dns分流确实进行了2次。对ssrp来说，它并不知道它转发的国内国外的dns上游是同一个。当它收到某个域名的上游返回的dns解析结果时，不管这个结果mosdns是分流到国内还是国外，ssrp只会按照自己的分流规则决定这个ip是否走代理。当然实际2种分流不一样的可能性不大。

所以我关注的点是，ssrp本身已经通过dnsmasq实现了dns的分流，mosdns是否有必要再做一次？如果有必要，好处是什么？

**Reply by DoveKi** [reply]
有个疑问openwrt上mosdns开启dns转发接管以后，系统自带dnsmasq本身的dns缓存需要关掉么
需要设置为0才会走mosdns的缓存吗

**Reply by yyysuo** [reply]
如果是dnsmasq转发，那dnsmasq并不是关掉了 ，是正常动作的，它的缓存当然也正常运行。mosdns的缓存也是正常运行的，每个环节都有缓存。

**Comment by qingtian110** [comment]
如果你完全是针对科学的话，mosdns确实没有任何必要，科学插件自带的分流和mosdns 如出一辙。

但是mosdns绝对不是为了科学上网而存在的，他只是刚好实现了与科学类同的分流功能罢了，它是一个独立的dns服务器，就好比有些人需要装个smartdns、AdGuardHome等这些dns插件来实现一些解析上的需求，mosdns 也是这个道理。

 mosdns 到底能干嘛？完全自定义任意的解析操作，它并不局限于地域的分流，它到底能做些什么 ，其实是没有答案的 https://irine-sistiana.gitbook.io/mosdns-wiki/mosdns/cha-jian-ji-qi-can-shu  mosdns自带的插件能实现无数种条件解析

它的好处是什么。性能好

**Reply by qingtian110** [reply]
> 如果你完全是针对科学的话，mosdns确实没有任何必要，科学插件自带的分流和mosdns 如出一辙。
> 
> 但是mosdns绝对不是为了科学上网而存在的，他只是刚好实现了与科学类同的分流功能罢了，它是一个独立的dns服务器，就好比有些人需要装个smartdns、AdGuardHome等这些dns插件来实现一些解析上的需求，mosdns 也是这个道理。
> 
> mosdns 到底能干嘛？完全自定义任意的解析操作，它并不局限于地域的分流，它到底能做些什么 ，其实是没有答案的 https://irine-sistiana.gitbook.io/mosdns-wiki/mosdns/cha-jian-ji-qi-can-shu mosdns自带的插件能实现无数种条件解析
> 
> 它的好处是什么。性能好

如果我第一次看到这个描述，我的评论绝对是 “吹什么吹，性能说好它就好了？”


所以、开源dns基准测试  https://github.com/DNS-OARC/dnsperf

**Reply by yyysuo** [reply]
> 如果你完全是针对科学的话，mosdns确实没有任何必要，科学插件自带的分流和mosdns 如出一辙。
> 
> 但是mosdns绝对不是为了科学上网而存在的，他只是刚好实现了与科学类同的分流功能罢了，它是一个独立的dns服务器，就好比有些人需要装个smartdns、AdGuardHome等这些dns插件来实现一些解析上的需求，mosdns 也是这个道理。
> 
> mosdns 到底能干嘛？完全自定义任意的解析操作，它并不局限于地域的分流，它到底能做些什么 ，其实是没有答案的 https://irine-sistiana.gitbook.io/mosdns-wiki/mosdns/cha-jian-ji-qi-can-shu mosdns自带的插件能实现无数种条件解析
> 
> 它的好处是什么。性能好

多谢解惑，跟surge一样了，surge其实也是个网络工具，但是应用最广的还是科学，mosdns是个好工具，但是网上大部分针对它与科学软件配合的教程，都是有点问题的。

**Reply by eskimokk** [reply]
你这个回答特别不好。一个产品是为了解决某个需求被发明出来的。
你说mosdns是为了什么被发明出来的，如果没有科学需求的话？

**Comment by JunShine** [comment]
有个疑惑,docker版的mosdns是不是无法写入ipset?理解中docker容器的概念,应该是容器与宿主机隔离了吧?mosdns设置写入ipset能写入宿主机的吗?

**Reply by mengxinz** [reply]
可以写入ipset,网络可以共用

**Comment by urlesistiana** [comment]
个人观点。科学软件如果自带 dns 分流，当然首选自带的分流策略。软件层面分流灵活，完善，简单。唯一缺点是性能比内核分流，ipset/nftabes之类的，差。但个人使用的话，不需要在意这点性能差距。除非遇到性能瓶颈了，比如有跑千兆以上的发烧需求。

mosdns 只能干一件事，就是在使用内核分流时优化 dns。它跟科学上网没联系。

> 但是mosdns绝对不是为了科学上网而存在的，他只是刚好实现了与科学类同的分流功能罢了，它是一个独立的dns服务器，就好比有些人需要装个smartdns、AdGuardHome等这些dns插件来实现一些解析上的需求，mosdns 也是这个道理。

100% 赞同

**Reply by BitbeyHub** [reply]
MosDNS比Passwall自带的DNS快多了，加载微信文章里的多图都不转圈了

**Reply by eskimokk** [reply]
那个事微信自己优化的，我没用mosdns之前就发现了。。。

**Comment by fmsonic** [comment]
我出现个问题，就是passwall+mosdns（luci），开启mosdns的转发，就无法科学上网了（或者开始能科学上网，第二天就不行了）

**Comment by likaia** [comment]
> 我出现个问题，就是passwall+mosdns（luci），开启mosdns的转发，就无法科学上网了（或者开始能科学上网，第二天就不行了）

你这个问题我也出现了，如果填了通过udp 请求dns，转发到mosdns的端口，就没法出墙了。

但是，我把mosdns 和 passwall 全部先关闭，然后清理passwall 的ipset ，dns转发用他自己的，填了1.1.1.1，然后启动。

最后再启动misdns，然后我奇迹的发现他走了mosdns，ping 谷歌的地址，延迟也降低了很多。

**Reply by fmsonic** [reply]
我的处理办法是passwall的dns设置1270.0.1:5335（就是指向mosdns），然后就是在mosdns把转发关掉就行了。结果就是远程dns指向mosdns，直连dns走路由自己的

**Reply by likaia** [reply]
我这样设置也可以了😂，但是有个新问题，paswall的分流节点失效了，他没办法分流了

**Comment by likaia** [comment]
> 我这样设置也可以了😂，但是有个新问题，paswall的分流节点失效了，他没办法分流了




> 我这样设置也可以了😂，但是有个新问题，paswall的分流节点失效了，他没办法分流了

这个问题我也解决了，是我这边的锅。

**Reply by fmsonic** [reply]
解决就好

**Reply by ruliuyan** [reply]
怎么解决的，出现一模一样的问题。不过经过我测试，发现是pw和mos启动先后的问题，passwall只能比mos后启动才能正常使用，如果mos重启，那pw就必须跟着重启才行。刚给mosdns和,uci两边都提了issue

**Reply by kokpk** [reply]
mosdns把转发关掉，paswall的分流节点失效。这是怎么解决的求指导

**Comment by guojiahao999** [comment]
有个问题 我passwall关闭了udp代理 mosdns会如何向远程dns发请求

**Reply by hadow** [reply]
passwall 关闭了udp代理，应该不会向mosdns 远程dns请求了，dnmasq 可以设置转发到mosdns

**Comment by xuy132** [comment]
![屏幕截图 2025-01-11 195004](https://github.com/user-attachments/assets/708025cc-d975-4984-a65e-891662bfc62d)
直接抄这个套娃吧！必须先启动modnns。
